src/main/java/com/google/sps/servlets/RecipeServlet.java

```diff
@@ -66,121 +78,132 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
+    if (newRecipe.metadata == null || newRecipe.content == null || newRecipe.metadata.title == null
+        || newRecipe.metadata.id == null) {
       response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
       return;
     }
 
     String token = request.getParameter("token");
-    FirebaseToken decodedToken = Auth.verifyIdToken(token);
-    if (decodedToken == null) {
-      response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
-      return;
-    }
-    if (!User.createdRecipe(decodedToken.getUid(), newRecipe.id)) {
-      response.setStatus(HttpServletResponse.SC_FORBIDDEN);
+    String uid = matchUser(token, newRecipe.metadata.id, response);
+    if (uid == null) {
       return;
     }
 
-    DocumentReference recipeRef = DBUtils.recipe(newRecipe.id);
-    ApiFuture future =
-        recipeRef.update(Recipe.TITLE_KEY, newRecipe.title, Recipe.CONTENT_KEY, newRecipe.content);
-    DBUtils.blockOnFuture(future);
+    db.editRecipeTitleContent(newRecipe.metadata.id, newRecipe.metadata.title, newRecipe.content);
   }
 
   @Override
   public void doDelete(HttpServletRequest request, HttpServletResponse response)
       throws IOException {
-    String recipeID = request.getParameter("recipeID");
-    if (recipeID == null) {
+    String recipeId = request.getParameter("recipeID");
+    if (recipeId == null) {
       response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
       return;
     }
 
     String token = request.getParameter("token");
-    FirebaseToken decodedToken = Auth.verifyIdToken(token);
-    if (decodedToken == null) {
-      response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
-      return;
-    }
-    if (!User.createdRecipe(decodedToken.getUid(), recipeID)) {
-      response.setStatus(HttpServletResponse.SC_FORBIDDEN);
+    String uid = matchUser(token, recipeId, response);
+    if (uid == null) {
       return;
     }
 
-    deleteComments(recipeID);
-    ApiFuture<WriteResult> writeResult = DBUtils.recipes().document(recipeID).delete();
-    DBUtils.blockOnFuture(writeResult);
+    db.deleteComments(recipeId);
+    db.deleteRecipe(recipeId);
   }
 
-  private String getRecipeList() {
-    Query query = DBUtils.recipes();
-    ApiFuture<QuerySnapshot> querySnapshotFuture = query.get();
-    ArrayList<Object> recipeList = new ArrayList<>();
-    QuerySnapshot querySnapshot = DBUtils.blockOnFuture(querySnapshotFuture);
-
-    if (querySnapshot == null) {
-      return null;
-    }
-
-    for (DocumentSnapshot document : querySnapshot.getDocuments()) {
-      recipeList.add(document.getData());
+  protected String getRecipeList(HttpServletRequest request, HttpServletResponse response) {
+    String creatorToken = request.getParameter("token");
+
+    String tagIDs[] = request.getParameterValues("tagIDs");
+    boolean isSavedRequest = Boolean.parseBoolean(request.getParameter("saved"));
+
+    boolean isTagQuery = !(tagIDs == null || tagIDs.length == 0 || tagIDs[0].equals("None"));
```

Comment:
it is easier to read if you if you do 
isTagQuery = tagIDs !=null && tagIDs.length > 0 && ...
same for next line.